FreeRTOS/Test/Target/tests/fr6/test.c

```diff
@@ -0,0 +1,199 @@
+
+static void prvTaskA(void *pvParameters);
+static void prvTaskB(void *pvParameters);
+static void prvTaskC(void *pvParameters);
+
+#if configNUMBER_OF_CORES != 2
+#error Require two cores be configured for FreeRTOS
+#endif
+
+bool testFailed = false;
+bool testPassed = false;
+int taskBState = 0;
+
+void test_fr6TASK_SWITCHED_IN(void) {
+  UBaseType_t idx, numTasksRunning;
+  TaskStatus_t taskStatus[16];
+  UBaseType_t taskStatusArraySize = 16;
+  unsigned long totalRunTime;
+  int coreIndex = 0;
+  SchedTraceLogRow *logRow;
+  int retcode = 0;
+
+  static int taskSwitchCount = 0;
+  static bool taskARan = false;
+  static bool taskBRan = false;
+  static bool taskCRan = false;
+
+  if (!(testPassed || testFailed))
+  {
+    bool taskARunning = false;
+
+    numTasksRunning = uxTaskGetSystemState((TaskStatus_t * const)&taskStatus, taskStatusArraySize, &totalRunTime);
+
+    for(idx = 0; idx < numTasksRunning; idx++)
+    {
+      if ((strcmp(taskStatus[idx].pcTaskName, "TaskA") == 0) && (taskStatus[idx].eCurrentState == eRunning))
+      {
+        taskARunning = true;
+        taskARan = true;
+        //sendReport("TaskA\n\0", 0);
+      }
+      if ((strcmp(taskStatus[idx].pcTaskName, "TaskB") == 0) && (taskStatus[idx].eCurrentState == eRunning))
+      {
+        taskBRan = true;
+        //sendReport("TaskB\n\0", 0);
+      }
+      if ((strcmp(taskStatus[idx].pcTaskName, "TaskC") == 0) && (taskStatus[idx].eCurrentState == eRunning))
+      {
+        taskCRan = true;
+        //sendReport("TaskC\n\0", 0);
+      }
+    }
+
+    if ((taskBState > 0) && taskCRan)
+    {
+      if (!(taskARunning && taskBRan))
+      {
+        testFailed = true;
+      }
+      else
+      {
+        testPassed = true;
+      }
+    }
+
+    taskSwitchCount++;
+    if (taskSwitchCount > 2048)
+    {
+      //sendReport("2k task swiches.\n\0", 0);
+      testFailed = true;
+    }
+  }
+}
+
+TaskHandle_t taskA;
+
+void setup_test_fr6_001(void) {
+  xTaskCreate(prvTaskA, "TaskA", configMINIMAL_STACK_SIZE, NULL,
+              mainTASK_A_PRIORITY, &taskA);
+  vTaskCoreAffinitySet(taskA, 0x2);
+
+  xTaskCreate(prvTaskB, "TaskB", configMINIMAL_STACK_SIZE, NULL,
+              mainTASK_B_PRIORITY, NULL);
+
+  xTaskCreate(prvTaskC, "TaskC", configMINIMAL_STACK_SIZE, NULL,
+              mainTASK_C_PRIORITY, NULL);
+}
+
+void setUp(void) {} /* Is run before every test, put unit init calls here. */
+void tearDown(void) {
+} /* Is run after every test, put unit clean-up calls here. */
+
+int main(void) {
+  initTestEnvironment();
+
+  setup_test_fr6_001();
+
+  vTaskStartScheduler();
+
+  /* should never reach here */
+  panic_unsupported();
+
+  return 0; // UNITY_END is unreachable via this path. a state machine and
+            // counter is used so that just one child task will call it
+            // instead.
+}
+
+static void prvTaskA(void *pvParameters) {
+  // wait with preemption disabled
+  vTaskPreemptionDisable(taskA);
```

Comment:
Ah, I see what you mean. I can give taskA the same priority as the others? I think with 3 tasks and 2 cores, the current scheduler may starve taskA otherwise given other experiments I've run.

Reply:
If task A has same priority with other two tasks, that it has 50% change not to preempt no matter what's setting on task.
So, I suggest to set task A with lowest priority so that it should be preempted (even it's preemptable).